### PR TITLE
DEV: implement intrinsic `pack`

### PR DESCRIPTION
**Pull request by Pranavchiku**
Towards #3457.

### CODE REVIEW

**Comment by certik** [comment]
The Windows failure is due to https://github.com/lfortran/lfortran/issues/3467, the other failure is probably due to unrelated CI reasons.

File: integration_tests/intrinsics_147.f90 [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  logical :: b
+  logical :: a(5)
+  real :: vector(7)
+  complex :: p(2)
+
+  vector = [5.0, 6.0, 3.0, 4.0, 0.0, 13.0, 29.0]
+  m = [1, 0, 0, 2, 99]
+  a = [.false., .false., .false., .false., .true.]
+
+  print *, pack(m, (/.false., .false., .false., .false., .true./), (/ 5, 6, 3, 4, 0, 13, 29 /))
```
Comment (certik) [inline integration_tests/intrinsics_147.f90:17]:
```suggestion
  print *, pack(m, [.false., .false., .false., .false., .true.], [ 5, 6, 3, 4, 0, 13, 29 ])
```

**Review by certik — APPROVED** [review]

**Comment by Pranavchiku** [comment]
The CI failure is same as that in master, it is because of #3467, will merge this, we can revert it any moment we wish.

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 2]
```diff
@@ -1725,6 +1727,340 @@ namespace MatMul {
...
+        }
+        int vector_rank = 0;
+        if (is_vector_present) {
+            vector_rank = extract_dimensions_from_ttype(type_vector, vector_dims);
+        }
+        if (array_rank != 1) {
+            append_error(diag, "`pack` accepts arrays of rank 1 only, provided an array "
+                "with rank, " + std::to_string(array_rank), array->base.loc);
+            return nullptr;
+        }
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_array_function_registry.h:1966]: This is incorrect, pack accepts any array, it returns an array of rank 1.

Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_array_function_registry.h:1966]: I am -1 on merging this.

Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_array_function_registry.h]: Implemented

**Comment by Pranavchiku** [comment]
This PR unblocks stdlib though there are a few cases for compiletime array for rank > 1 `pack(reshape([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], [3, 4]), .true.)` that needs work but it is not needed for `stdlib` at moment,

File: integration_tests/CMakeLists.txt [hunk 3]
```diff
@@ -367,9 +367,9 @@ RUN(NAME arrays_02_size LABELS gfortran llvm llvmStackArray wasm fortran)
 RUN(NAME arrays_03_size LABELS gfortran llvm llvmStackArray fortran)
 RUN(NAME matrix_01_transpose LABELS gfortran llvm)
 RUN(NAME matrix_02_matmul LABELS gfortran)
-RUN(NAME array_01_pack LABELS gfortran fortran)
+RUN(NAME array_01_pack LABELS gfortran llvm)
```
Comment (czgdp1807) [inline integration_tests/CMakeLists.txt:370]: Why remove ``fortran`` from tests?

Comment (Pranavchiku) [reply integration_tests/CMakeLists.txt:370]: Previously it used `ArrayPack` node and that has implementation in fortran backend, now we use `IntrinsicArrayFunction` and `Pack` does not have support for it. I have #3416 opened for these issues.

Comment (Pranavchiku) [reply integration_tests/CMakeLists.txt:370]: If rest all is good, I will prefer to merge this before beginning work on `unpack`. 
cc @czgdp1807

Comment (czgdp1807) [reply integration_tests/CMakeLists.txt:370]: > Previously it used ArrayPack node and that has implementation in fortran backend

Just hook up that implementation with `IntrinsicArrayFunction`. It is easy to do.

Comment (czgdp1807) [reply integration_tests/CMakeLists.txt:370]: > If rest all is good, I will prefer to merge this before beginning work on `unpack`.

Only @certik has the access to do it.

Comment (Pranavchiku) [reply integration_tests/CMakeLists.txt]: > Just hook up that implementation with IntrinsicArrayFunction. It is easy to do.

Done.

**Comment by Pranavchiku** [comment]
I have changes for `unpack` ready, will open PR once this is merged.

**Comment by czgdp1807** [comment]
I think reference tests need to be updated.

**Comment by Pranavchiku** [comment]
I did a clean build after rebasing manually, tests works for me.

```console
% ./run_tests.py 
[..]
warnings/open_notimplemented_kwarg1.f90 * asr ✓ 
errors/open_invalid_kwarg1.f90 * asr ✓ 
errors/substring_endidx.f90 * asr ✓ 
shiftl1.f90 * asr ✓ 
errors/substring_startidx.f90 * asr ✓ 
errors/fixed_number_of_args.f90 * asr ✓ 
errors/close_invalid_kwarg1.f90 * asr ✓ 
errors/compare_01.f90 * asr ✓ 
errors/flush_invalid_kwarg1.f90 * asr ✓ 
errors/runtime_stacktrace_01.f90 * run_dbg ✓ 
TESTS PASSED
```

**Comment by czgdp1807** [comment]
Yeah. I think that's flaky. In fact that particular reference test might contain some hidden bugs. May be the C++ implementation uses something which doesn't preserve the order and that's why we face the error with that reference test. Anyways, its tested with LLVM, so we can remove it from reference tests.

**Comment by certik** [comment]
> Yeah. I think that's flaky. In fact that particular reference test might contain some hidden bugs.

Can you open up an issue for this bug / flaky test? So that we don't forget to investigate later and fix it.

**Pull request merged**